File: src/solves/day02.rs
use std::fs;

pub fn part1() {
    let input = fs::read_to_string("./inputs/input02").expect("no file");
    let reports: Vec<Vec<i32>> = input
        .lines()
        .map(|line| {
            line.split_whitespace()
                .map(|n| n.parse::<i32>().unwrap())
                .collect()
        })
        .collect();

    let sum = reports
        .iter()
        .filter(|r| {
            let mut inc = None;
            r.windows(2).all(|l| {
                if (l[1] - l[0]).abs() > 3 {
                    return false;
                }

                if l[1] > l[0] {
                    if let Some(false) = inc {
                        return false;
                    }
                    inc = Some(true);
                } else if l[1] < l[0] {
                    if let Some(true) = inc {
                        return false;
                    }
                    inc = Some(false);
                } else {
                    return false;
                }

                true
            })
        })
        .count();

    println!("{}", sum);
}

pub fn part2() {
    let input = fs::read_to_string("./inputs/input02").expect("no file");
    let reports: Vec<Vec<i32>> = input
        .lines()
        .map(|line| {
            line.split_whitespace()
                .map(|n| n.parse::<i32>().unwrap())
                .collect()
        })
        .collect();

    let sum = reports
        .iter()
        .filter(|r| {
            let mut inc = None;
            let mut tol = false;
            r.windows(2).all(|l| {
                if (l[1] - l[0]).abs() > 3 {
                    if tol == true {
                        return false;
                    }
                    tol = true;
                }
                if l[1] > l[0] {
                    if let Some(false) = inc {
                        if tol == true {
                            return false;
                        }
                        tol = true;
                    }
                    inc = Some(true);
                } else if l[1] < l[0] {
                    if let Some(true) = inc {
                        if tol == true {
                            return false;
                        }
                        tol = true;
                    }
                    inc = Some(false);
                } else {
                    if tol == true {
                        return false;
                    }
                    tol = true;
                }
                true
            })
        })
        .count();

    println!("{}", sum);
}

File: solves/day02.py
def check1(r):
    inc = None
    for i in range(len(r) - 1):
        if abs(r[i + 1] - r[i]) > 3:
            return False
        if r[i + 1] > r[i]:
            if inc is False:
                return False
            inc = True
        elif r[i + 1] < r[i]:
            if inc is True:
                return False
            inc = False
        else:
            return False
    return True

def check2(r):
    inc = None
    tol = False
    for i in range(len(r) - 1):
        if abs(r[i + 1] - r[i]) > 3:
            if tol:
                return False
            tol = True
        if r[i + 1] > r[i]:
            if inc is False:
                if tol:
                    return False
                tol = True
            inc = True
        elif r[i + 1] < r[i]:
            if inc is True:
                if tol:
                    return False
                tol = True
            inc = False
        else:
            return False
    return True

def solve1():
    with open("./inputs/input02", "r") as file:
        lines = file.readlines()

    reports = [[int(num) for num in line.split()] for line in lines]

    res = sum(1 for r in reports if check1(r))

    print(res)

def solve2():
    with open("./inputs/input02", "r") as file:
        lines = file.readlines()

    reports = [[int(num) for num in line.split()] for line in lines]

    res = sum(1 for r in reports if check2(r))

    print(res)

